### ISSUE TITLE
Wrap widget sometimes clips its contents

### ISSUE THREAD

**Issue by ameiyil** [flutter/flutter#11488]
## Steps to Reproduce
I am populating a Wrap widget with chips created from a list of values.
I'm noticing sometimes the Wrap widget is off by a few pixels and places a chip which then is clipped on one side.

All of this is contained in a StatefulWidget in which the Wrap widget and a TextField widget reside. A list of values is maintained by the widget, and the list is mutated as the user types and submits content from a text field. 

I have attached a screenshot and the snippet of code for the widget.
<img width="338" alt="screen shot 2017-08-02 at 10 19 00 am" src="https://user-images.githubusercontent.com/28424281/28885826-0cee78a6-776c-11e7-800c-037cc3e9dd38.png">

## Code Snippet
I purposely constrained the width to smaller than the available space and gave my container a background color to make sure that it wasn't being occluded or constrained by it's parent or anything.

This is the code snippet that produces the wrap widget in the screenshot.
```
        new Container(
          width: 120.0, 
          color: Colors.blueGrey,
          constraints: new BoxConstraints(
            maxHeight: 150.0, 
          ),
          child: new SingleChildScrollView(
            reverse: true,
            child: new Wrap(
              spacing: 8.0,
              runSpacing: 4.0,
              children: _chipsContent.map((String text) {
                return new Chip(
                  label: new Text(text),
                  onDeleted: () {
                    setState(() {
                      _chipsContent.remove(text);
                    });
                  },
                );
              }).toList(),
            ),
         ),
```

## Logs
Will include as needed

## Flutter Doctor
[✓] Flutter (on Linux, locale en_US.UTF-8, channel alpha)
    • Flutter at <my flutter dir>
    • Framework revision d36e2f6191 (6 weeks ago), 2017-06-21 15:09:10 -0700
    • Engine revision b0dee695ec
    • Tools Dart version 1.24.0-dev.6.7

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.0)
    • Android SDK at <my android sdk dir>
    • Platform android-26, build-tools 26.0.0
    • Java binary at: /usr/local/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio (version 2.3)
    • Android Studio at /usr/local/android-studio
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.1)
    • Flutter plugin version 15.1
    • Dart plugin version 171.4424

[✓] IntelliJ IDEA Community Edition (version 2017.1)
    • Flutter plugin version 15.1
    • Dart plugin version 171.4424

[✓] Connected devices
    • None

**Comment by amardeep** [comment]
I am also facing a similar issue. Attached code snippet and screenshot

![image](https://user-images.githubusercontent.com/86928/30108003-34369b28-931e-11e7-97bc-3736530b3b3c.png)

As you can see in the image, the word "wholesale" is cut off.

```
List<String> keywords;
Widget build(context) {
    var theme = Theme.of(context);
    var textTheme = theme.textTheme;
    

    return new Row(children: [
      new Expanded(
          child: new Container(
              padding: new EdgeInsets.all(16.0),
              decoration: new BoxDecoration(
                  border: new Border(
                      bottom: new BorderSide(color: theme.dividerColor))),
              child: new Wrap(
                  spacing: 18.0,
                  children: keywords
                      .map((k) => new GestureDetector(
                          onTap: () {
                            print('tapped $k');
                          },
                          child: new Text(
                            k,
                            style: textTheme.body2
                                .copyWith(fontSize: 12.0, color: Colors.blue),
                          )))
                      .toList())))
    ]);
  }
```

**Comment by jklint-g** [comment]
We ran into this bug for mulligan as well.

**Comment by Hixie** [comment]
This is on our high-priority list, we now need to find someone to work on it.

**Comment by jonahwilliams** [comment]
I started taking a look at the examples above this morning.  I can reproduce the second example on a pixel 1, but the first _seems_ to be rendering okay.  @ameiyil If you have any more context to reproducing the case with the Chips that would be great,

<img src="https://user-images.githubusercontent.com/8975114/37106815-955d6d8e-21e7-11e8-8222-eb44b7d1d701.png " width="40%">.

<img src="https://user-images.githubusercontent.com/8975114/37106823-998b2e00-21e7-11e8-8cf0-e4cb1d6abbc8.png" width="40%">.

**Comment by Hixie** [comment]
Maybe it happens at particular widths? e.g. when it would only overflow by 3 pixels or some such?

**Comment by jonahwilliams** [comment]
So, problem might be as simple as `RenderWrap.performLayout()` checks if the size of the child plus the main axis extend exceeds the main axis limit: `runMainAxisExtent + childMainAxisExtent > maxAxisLimit`. This calculation doesn't involve the `spacing`, however, so the child could overflow from 0 to `spacing` dp.  

Still confirming that this is the case

**Comment by jonahwilliams** [comment]
I believe this issue is fixed by #15256, as I didn't encounter any other issues in the Wrap widget.
